scwidgets/converters.py

```diff
@@ -0,0 +1,77 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+# Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
+# @file
+# @author Matthew Andrew
+
+import scipp as sc
+import pathlib
+
+scipp_object_mutable = (sc.DataArray, sc.Dataset, sc.Variable,
+                        sc.DataArrayView, sc.DatasetView, sc.VariableView)
+
+scipp_object = scipp_object_mutable + (
+    sc.DataArrayConstView, sc.DatasetConstView, sc.VariableConstView)
+
+
+def filepath_converter(filepath, data_directory=pathlib.Path.cwd()):
+    """
+    Checks the validity of the filepath provided,
+    returns it is valid. Otherwise tries to form a valid
+    filepath assuming a filename has been provided instead.
+
+    Throws if file not found.
+    """
+    path = pathlib.Path(filepath)
+
+    if path.is_file():
+        return filepath
+
+    path = data_directory / path
+
+    if path.is_file():
+        return str(path)
+
+    raise ValueError(f'Filepath {filepath} was not found.')
+
+
+def scope_converter(object_name, scope):
+    """
+    Attempts to convert the string input into an object by
+    assuming it names an object in the provided scope.
+
+    Throws if not object of correct name found.
+    """
+    if object_name in scope:
+        return scope[object_name]
+
+    raise ValueError(
+        f'Object of name {object_name} does not exist within scope.')
+
+
+def typed_scope_converter(object_name, allowed_types, scope):
+    """
+    Attempts to convert the string input into an scipp object by
+    assuming it names an object in the provided scope.
+
+    Throws if object is not found or not of an allowed type.
+    """
+    object = scope_converter(object_name, scope)
+
+    if not isinstance(object, allowed_types):
+        valid_types = ','.join(str(item.__name__) for item in allowed_types)
+        raise ValueError(
+            f'{object_name} of invalid type. Valid types are: {valid_types}')
+
+    return object
+
+
+def string_allowed_values_converter(input, allowed_values):
```

Comment:
`dropdown_input` for a more concise name?

Reply:
This doesn't effect which values appear in the dropdown menu. It merely validates that the value inputted is from an allowed selection. The drop down menu options are controlled by a separate dictionary in the widget construction. The current design is that for each input there are four things which can be varied: hinting text, conversion function, validation, drop-down options. 

These are controlled by:
  * conversion and validation are controlled by one function which throws if the validation fails and otherwise returns the converted object.
  * hinting-text is by default the input name but can be modified via the description dictionary
  * drop-down menu items are controlled by the options dictionary.

From a design perspective I think the outstanding question is which of these options we want to include on the first pass and what is the best way to set them for example:
  * Do we want to bundle conversion and validation or do we want to have a mandatory conversion dict and an optional validation one in the same way as descriptions/options?
 * Do we want to bundle the valdation and options input instead such that where a validator is supplied drop-down options are populated if possible?

Reply:
> It merely validates that the value inputted is from an allowed selection. The drop down menu options are controlled by a separate dictionary in the widget construction.
This feels like you need to do two things to achieve one objective?

Reply:
> The current design is that for each input there are four things which can be varied: hinting text, conversion function, validation, drop-down options.

- Is drow-down options just a special case of the "type" of the input field. Can it be configured to, e.g., accept only numbers, or as a spin-box?
- Conversion and validation seem to be "the same" for me? Can we achive this by potentially chaining arbitrary functions? Basically we just want to call `process(func1(func2(input))`, where the number of `funcs` is aribtrary and `func1` could be conversion, `func2` could be validation, or vice versa, ...

Reply:
Yes that is the direction my thoughts on your comments are leading me in. I think a better input layout would be

`ProcessWidget(callable, name, converters, validators={}, descriptions={})`

So you bundle the validator and options and separate the converters and validators?

Reply:
Can we make a class that serves as an input?

```python
input1 = ComboInput(options={'a','b'}, process=my_validator, tooltip='xxx', name='ab')
input2 = NumberInput(process=lambda x: my_validator(my_converter(x))) # also support some way of chaining
ProcessWidget(callable, name, inputs=[input1, input2])
```

Reply:
Yes I'll give the classes method a go.